File: back/src/main/java/com/funstat/MainController.java
package com.funstat;

import com.funstat.domain.Annotations;
import com.funstat.domain.Ohlc;
import com.funstat.domain.TimePoint;
import com.funstat.finam.InstrId;
import com.funstat.ohlc.Metadata;
import com.funstat.store.MdStorage;
import com.funstat.store.MdStorageImpl;
import org.springframework.context.annotation.Bean;
import org.springframework.web.bind.annotation.*;
import springfox.documentation.builders.PathSelectors;
import springfox.documentation.builders.RequestHandlerSelectors;
import springfox.documentation.spi.DocumentationType;
import springfox.documentation.spring.web.plugins.Docket;

import javax.annotation.PostConstruct;
import javax.validation.Valid;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;


@RestController
@CrossOrigin(origins = "*")
public class MainController {

    MdStorage storage = new MdStorageImpl("/ddisk/globaldatabase/md");

    @PostConstruct
    void onStart(){
        storage.updateSymbolsMeta();
    }

    List<Metadata> allMetas = new ArrayList<>();

    @GetMapping(value = "/metas")
    public List<Metadata> loadAllMetas(){
        return allMetas;
    }

    @PutMapping("/put_meta")
    public void addMetadata(@Valid @RequestParam Metadata metadata){
        allMetas.add(metadata);
    }

    @RequestMapping(value="/instruments", method=RequestMethod.GET)
    public Collection<InstrId> instruments(){
        return storage.getMeta();
    }

    @PostMapping("/get_ohlcs")
    public Collection<Ohlc> getOhlcs(@RequestBody @Valid InstrId instrId, String interval){
        return storage.read(instrId, interval);
    }

    @PostMapping("/get_annotations")
    public Annotations getAnnotations(@RequestBody @Valid InstrId instrId, String interval){
        List<Ohlc> ohlcs = storage.read(instrId, interval);
        return AnnotationCreator.createAnnotations(ohlcs);
    }

    @PostMapping("/get_series")
    public Map<String,Collection<TimePoint>> getSeries(@RequestBody @Valid InstrId[] codes, String interval) {
        Map<String,Collection<TimePoint>> mm = new HashMap<>();
        Arrays.stream(codes).forEach(c->{
            List<TimePoint> lst = storage.read(c,interval).stream().map(oh -> new TimePoint(oh.dateTime, oh.close)).collect(Collectors.toList());
            Collections.sort(lst);
            mm.put(c.code, lst);
        });
        return mm;
    }

    @Bean
    public Docket customImplementation() {
     return new Docket(DocumentationType.SWAGGER_2)
                 //here is the line that you need
                .directModelSubstitute(LocalDateTime.class, String.class)
                .directModelSubstitute(LocalDate.class, String.class)
                .select()
                .apis(RequestHandlerSelectors.any())

                .paths(PathSelectors.regex("/.*"))
                .build()
                .pathMapping("/");    }
}

File: back/src/main/java/com/funstat/MainController.kt
package com.funstat

import com.funstat.domain.Annotations
import com.funstat.domain.StringWrap
import com.funstat.domain.TimePoint
import firelib.core.domain.InstrId
import firelib.core.domain.Interval
import firelib.core.domain.Ohlc
import firelib.core.domain.sourceEnum
import firelib.core.misc.atUtc
import firelib.core.store.*
import firelib.vantage.VantageDownloader
import org.slf4j.LoggerFactory
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*


class MainController {


    val log = LoggerFactory.getLogger(javaClass)

    internal var storage: MdStorage = MdStorageImpl(GlobalConstants.mdFolder.toString())

    internal var allMetas: MutableList<Metadata> = ArrayList()

    val noteBooksDir = Paths.get(System.getProperty("notebooksDir") ?: "/home/ivan/projects/fbackend/market_research/published" )

    init {
        log.info("notebooks dir is " + noteBooksDir)
    }

    fun loadStaticPages(): List<String> {
        return noteBooksDir.toFile()
                .list { f, name -> name.endsWith("html") }
                .map { it.replace(".html", "") }
    }

    fun loadAllMetas(): List<Metadata> {
        return allMetas
    }


    fun loadHtmContent(file: String): StringWrap {
        return StringWrap(String(Files.readAllBytes(noteBooksDir.resolve("$file.html"))))
    }


    fun addMetadata(metadata: Metadata) {
        allMetas.add(metadata)
    }

    fun instruments(): Collection<InstrId> {
        return storage.meta().filter { s -> s.sourceEnum() != VantageDownloader.SOURCE }
    }

    fun timeframes(): List<String> {
        return Interval.values().map { it.name }
    }


    fun getOhlcs(instrId: InstrId, interval: Interval): Collection<Ohlc> {
        return storage.read(instrId, interval, interval)
    }

    fun getAnnotations(instrId: InstrId, interval: Interval): Annotations {
        val ohlcs = storage.read(instrId, interval, interval)
        return AnnotationCreator.createAnnotations(ohlcs)
    }

    fun getSeries(codes: Array<InstrId>, interval: Interval): Map<String, Collection<TimePoint>> {
        val mm = HashMap<String, Collection<TimePoint>>()
        Arrays.stream(codes).forEach { c ->
            mm[c.code] = storage.read(c, interval, interval).map { oh -> TimePoint(oh.endTime.atUtc(), oh.close) }.sorted()
        }
        return mm
    }

}